onnx/shape_inference/implementation.cc

```diff
@@ -588,10 +588,12 @@ class ShapeInferenceImplBase {
     // Create a temporary initializer value map
     for (int i = 0; i < num_actual_inputs && i < num_func_inputs; ++i) {
       const TypeProto* type = ctx.getInputType(i);
-      if (type->value_case() == TypeProto::kTensorType && ctx.getInputData(i) != nullptr) {
-        input_data_by_name[func_proto.input().Get(i)] = ctx.getInputData(i);
-      } else if (type->value_case() == TypeProto::kSparseTensorType && ctx.getInputSparseData(i) != nullptr) {
-        input_sparse_data_by_name[func_proto.input().Get(i)] = ctx.getInputSparseData(i);
+      if (type != nullptr) {
```

Comment:
What if the optional parameter is the first one and the second one is defined? What about the line 876 in that case?

Reply:
It looks fine to me. I also don't see the connection between this line and line 876. Line 876 is returning the type for a particular input, and the contract is to return nullptr in the case of missing optional inputs. This line is part of the pre-processing to create known constant-values of inputs to help shape-inference. This line change is just fixing a bug that it wasn't handling optional inputs correctly.

Reply:
Added test-case with a missing non-trailing optional parameter.